### PR TITLE
Fix NaN handling in is_sign_negative/positive

### PR DESCRIPTION
**Pull request by nagisa**
This would be my proposed fix for the #42425 provided we decide it is indeed a problem.

Note this would technically be a breaking change to a stable API. We might want to consider deprecating these methods and adding new ones.

### CODE REVIEW

File: src/libcore/num/f32.rs [hunk 1]
```diff
@@ -205,18 +205,25 @@ impl Float for f32 {
         }
     }
 
-    /// Returns `true` if `self` is positive, including `+0.0` and
-    /// `Float::infinity()`.
+    /// Returns `true` if and only if `self` is has a positive sign, including `-0.0`, `NaN`s with
```
Comment (retep998) [inline src/libcore/num/f32.rs]: `is has`?

Comment (Thiez) [inline src/libcore/num/f32.rs]: I'm more surprised that it returns `true` for "`-0.0`, `NaN`s with negative sign, and negative infinity". I would not expect those cases to have a positive sign.

**Comment by natboehm** [comment]
```
[01:11:11] failures:
[01:11:11]     f32::tests::test_is_sign_positive
[01:11:11]     f32::tests::test_nan
[01:11:11]     f64::tests::test_is_sign_positive
[01:11:11]     f64::tests::test_nan
```

It looks like some tests are failing.

**Comment by nagisa** [comment]
This is not supposed to be in a landable state until we reach any concensus anyway.

**Comment by arielb1** [comment]
What's the status of this @nagisa? Just checking in to make sure it isn't getting lost.

**Comment by alexcrichton** [comment]
Discussed during libs triage the conclusion was that this seems good to land! @nagisa want to fix up the tests and I'll r+?

**Comment by alexcrichton** [comment]
Looks like there's some Travis errors?

```
[01:11:14] failures:
[01:11:14] 
[01:11:14] ---- f32.rs - f32::f32::is_sign_negative (line 388) stdout ----
[01:11:14] 	thread 'rustc' panicked at 'test executable failed:
[01:11:14] 
[01:11:14] thread 'main' panicked at 'assertion failed: !nan.is_sign_positive() && !nan.is_sign_negative()', <anon>:13
[01:11:14] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:11:14] 
[01:11:14] ', /checkout/src/librustdoc/test.rs:318
[01:11:14] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:11:14] 
[01:11:14] ---- f32.rs - f32::f32::is_sign_positive (line 369) stdout ----
[01:11:14] 	thread 'rustc' panicked at 'test executable failed:
[01:11:14] 
[01:11:14] thread 'main' panicked at 'assertion failed: !nan.is_sign_positive() && !nan.is_sign_negative()', <anon>:13
[01:11:14] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:11:14] 
[01:11:14] ', /checkout/src/librustdoc/test.rs:318
[01:11:14] 
[01:11:14] ---- f64.rs - f64::f64::is_sign_negative (line 332) stdout ----
[01:11:14] 	thread 'rustc' panicked at 'test executable failed:
[01:11:14] 
[01:11:14] thread 'main' panicked at 'assertion failed: !nan.is_sign_positive() && !nan.is_sign_negative()', <anon>:14
[01:11:14] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:11:14] 
[01:11:14] ', /checkout/src/librustdoc/test.rs:318
[01:11:14] 
[01:11:14] ---- f64.rs - f64::f64::is_sign_positive (line 307) stdout ----
[01:11:14] 	thread 'rustc' panicked at 'test executable failed:
[01:11:14] 
[01:11:14] thread 'main' panicked at 'assertion failed: !nan.is_sign_positive() && !nan.is_sign_negative()', <anon>:14
[01:11:14] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:11:14] 
[01:11:14] ', /checkout/src/librustdoc/test.rs:318
[01:11:14] 
[01:11:14] 
[01:11:14] failures:
[01:11:14]     f32.rs - f32::f32::is_sign_negative (line 388)
[01:11:14]     f32.rs - f32::f32::is_sign_positive (line 369)
[01:11:14]     f64.rs - f64::f64::is_sign_negative (line 332)
[01:11:14]     f64.rs - f64::f64::is_sign_positive (line 307)
[01:11:14] 
[01:11:14] test result: FAILED. 820 passed; 4 failed; 22 ignored; 0 measured; 0 filtered out
```

**Comment by arielb1** [comment]
@nagisa is this ready for review?

**Comment by nagisa** [comment]
@alexcrichton tests pass.

**Pull request merged**